File: android/app/src/main/java/com/openrung/net/GeoIpClient.kt
package com.openrung.net

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

data class ClientGeoInfo(
    val ip: String,
    val country: String,
    val countryCode: String,
    val city: String,
    val asn: String,
    val isp: String,
    val organization: String,
    val latitude: Double = 0.0,
    val longitude: Double = 0.0,
) {
    fun telemetryAttributes(): Map<String, String> = mapOf(
        "client_ip" to ip,
        "country" to country,
        "country_code" to countryCode,
        "city" to city,
        "asn" to asn,
        "isp" to isp,
        "organization" to organization,
    ).filterValues { it.isNotBlank() }

    /** Human-readable location such as "Austin, United States", or "" when unknown. */
    fun locationLabel(): String = listOf(city, country).filter { it.isNotBlank() }.joinToString(", ")
}

@Serializable
private data class GeoIpResponse(
    val ip: String = "",
    val success: Boolean = false,
    val country: String = "",
    @SerialName("country_code")
    val countryCode: String = "",
    val city: String = "",
    val latitude: Double = 0.0,
    val longitude: Double = 0.0,
    val connection: GeoIpConnection = GeoIpConnection(),
)

@Serializable
private data class GeoIpConnection(
    val asn: Long = 0,
    val org: String = "",
    val isp: String = "",
)

class GeoIpClient(
    private val endpoint: String = DEFAULT_ENDPOINT,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    /**
     * Looks up geo info for [ip], or for the caller's own public IP when [ip] is null/blank.
     */
    suspend fun lookup(ip: String? = null): ClientGeoInfo = withContext(Dispatchers.IO) {
        val target = if (ip.isNullOrBlank()) endpoint else endpoint.trimEnd('/') + "/" + ip.trim()
        val connection = (URL(target).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 4_000
            readTimeout = 4_000
            useCaches = false
            setRequestProperty("Accept", "application/json")
        }
        try {
            val status = connection.responseCode
            if (status !in 200..299) throw IOException("geo IP HTTP $status")
            decode(connection.inputStream.bufferedReader().use { it.readText() })
        } finally {
            connection.disconnect()
        }
    }

    internal fun decode(body: String): ClientGeoInfo {
        val response = json.decodeFromString<GeoIpResponse>(body)
        if (!response.success || response.ip.isBlank()) throw IOException("geo IP lookup failed")
        return ClientGeoInfo(
            ip = response.ip,
            country = response.country,
            countryCode = response.countryCode,
            city = response.city,
            asn = response.connection.asn.takeIf { it > 0 }?.let { "AS$it" }.orEmpty(),
            isp = response.connection.isp,
            organization = response.connection.org,
            latitude = response.latitude,
            longitude = response.longitude,
        )
    }

    companion object {
        const val DEFAULT_ENDPOINT = "https://ipwho.is/"
    }
}

File: src/net/geoIpClient.ts
/**
 * GeoIP lookup via https://ipwho.is/, ported from the production `net/GeoIpClient.kt`.
 * Production uses 4s connect + 4s read timeouts; RN `fetch` gets a single 4s overall deadline.
 */

export const DEFAULT_ENDPOINT = 'https://ipwho.is/';

const REQUEST_TIMEOUT_MS = 4_000;

export interface ClientGeoInfo {
  ip: string;
  country: string;
  countryCode: string;
  city: string;
  asn: string;
  isp: string;
  organization: string;
  latitude: number;
  longitude: number;
}

/** Telemetry attribute map (blank values filtered), as in production. */
export function telemetryAttributes(geo: ClientGeoInfo): Record<string, string> {
  const entries: Array<[string, string]> = [
    ['client_ip', geo.ip],
    ['country', geo.country],
    ['country_code', geo.countryCode],
    ['city', geo.city],
    ['asn', geo.asn],
    ['isp', geo.isp],
    ['organization', geo.organization],
  ];
  const result: Record<string, string> = {};
  for (const [key, value] of entries) {
    if (value.trim().length > 0) {
      result[key] = value;
    }
  }
  return result;
}

/** Human-readable location such as "Austin, United States", or "" when unknown. */
export function locationLabel(geo: ClientGeoInfo): string {
  return [geo.city, geo.country].filter(part => part.trim().length > 0).join(', ');
}

function asString(value: unknown): string {
  return typeof value === 'string' ? value : '';
}

function asNumber(value: unknown): number {
  return typeof value === 'number' && Number.isFinite(value) ? value : 0;
}

/** Decodes an ipwho.is response body (unknown keys ignored; `!success || blank ip` is an error). */
export function decodeGeoIpResponse(body: string): ClientGeoInfo {
  const parsed: unknown = JSON.parse(body);
  const record = (typeof parsed === 'object' && parsed !== null ? parsed : {}) as Record<
    string,
    unknown
  >;
  const success = record.success === true;
  const ip = asString(record.ip);
  if (!success || ip.trim().length === 0) {
    throw new Error('geo IP lookup failed');
  }
  const connection = (typeof record.connection === 'object' && record.connection !== null
    ? record.connection
    : {}) as Record<string, unknown>;
  const asn = asNumber(connection.asn);
  return {
    ip,
    country: asString(record.country),
    countryCode: asString(record.country_code),
    city: asString(record.city),
    asn: asn > 0 ? `AS${asn}` : '',
    isp: asString(connection.isp),
    organization: asString(connection.org),
    latitude: asNumber(record.latitude),
    longitude: asNumber(record.longitude),
  };
}

/**
 * Looks up geo info for `ip`, or for the caller's own public IP when `ip` is null/blank.
 */
export async function lookup(
  ip: string | null = null,
  endpoint: string = DEFAULT_ENDPOINT,
): Promise<ClientGeoInfo> {
  const target =
    ip === null || ip.trim().length === 0
      ? endpoint
      : endpoint.replace(/\/+$/, '') + '/' + ip.trim();

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
  try {
    const response = await fetch(target, {
      method: 'GET',
      headers: { Accept: 'application/json' },
      signal: controller.signal,
    });
    if (response.status < 200 || response.status > 299) {
      throw new Error(`geo IP HTTP ${response.status}`);
    }
    return decodeGeoIpResponse(await response.text());
  } finally {
    clearTimeout(timer);
  }
}
